File: ios/VocaPhoneShared/TranscriptStyler.swift
import Foundation

/// Local equivalent of the gateway's presentation-only transcript styles.
/// Words are never added, removed, or substituted; only case, spacing, and
/// sentence punctuation may change.
///
/// Dropping a filler or inserting a missing sentence break would both break
/// that contract, which is why they are ``TranscriptRepair``'s job and run
/// before this stage under a switch of their own.
enum TranscriptStyler {
    static func apply(
        _ text: String?,
        style: WritingStyle,
        language: String = "auto"
    ) -> String {
        let source = text ?? ""
        if style == .raw { return source.trimmingCharacters(in: .whitespacesAndNewlines) }
        guard !source.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else { return "" }

        let punctuation = SentencePunctuation.resolve(language: language, text: source)
        let spans = ProtectedSpans.mask(source)
        let normalized = normalizeSentenceTerminators(
            normalizeSpacing(spans.text),
            punctuation: punctuation
        )
        // Whisper and Parakeet Title-Case content words. Flatten those before
        // Clean/Formal/Casual so mid-sentence capitals are not left as-is.
        // Mixed-case names and ALL-CAPS acronyms stay.
        let flattened: String
        switch style {
        case .raw, .veryCasual:
            flattened = normalized
        default:
            flattened = flattenModelCaps(normalized)
        }
        let result: String
        switch style {
        case .raw:
            result = normalized
        case .clean:
            result = ensureTerminator(flattened, punctuation: punctuation)
        case .formal:
            result = ensureTerminator(
                capitalizeSentenceStarts(flattened, punctuation: punctuation),
                punctuation: punctuation
            )
        case .casual:
            result = casual(flattened, punctuation: punctuation)
        case .veryCasual:
            result = veryCasual(segments(normalized, punctuation: punctuation), punctuation: punctuation)
        case .excited:
            result = excited(segments(flattened, punctuation: punctuation), punctuation: punctuation)
        }
        let lowered = style == .veryCasual
            ? ProtectedSpans.mapOutsidePlaceholders(result) { $0.lowercased() }
            : result
        return spans.restore(lowered)
    }

    private static func normalizeSpacing(_ text: String) -> String {
        text
            .replacingOccurrences(of: "\\s+", with: " ", options: .regularExpression)
            .trimmingCharacters(in: .whitespacesAndNewlines)
            .replacingOccurrences(
                of: "\\s+([.!?。！？।۔،,;:])",
                with: "$1",
                options: .regularExpression
            )
    }

    /// A model often emits an ASCII full stop even when it correctly decoded
    /// Hindi text. Once the script is known, canonicalize sentence boundaries
    /// while leaving masked URLs, decimals, abbreviations, and ellipses intact.
    private static func normalizeSentenceTerminators(
        _ text: String,
        punctuation: SentencePunctuation
    ) -> String {
        guard punctuation.terminator == "।" else { return text }
        let characters = Array(text)
        var result = ""
        result.reserveCapacity(text.count)
        for index in characters.indices {
            let character = characters[index]
            guard character == "." else {
                result.append(character)
                continue
            }
            let previous = index > characters.startIndex ? characters[index - 1] : nil
            let nextIndex = index + 1
            let next = nextIndex < characters.endIndex ? characters[nextIndex] : nil
            let isEllipsis = previous == "." || next == "."
            let isSentenceBoundary = next == nil || next?.isWhitespace == true
            result.append(isSentenceBoundary && !isEllipsis ? "।" : character)
        }
        return result
    }

    private static func ensureTerminator(
        _ text: String,
        punctuation: SentencePunctuation
    ) -> String {
        guard !text.isEmpty, !punctuation.terminator.isEmpty else { return text }
        guard let last = text.last, !punctuation.terminators.contains(last) else { return text }
        return text + punctuation.terminator
    }

    /// Drop Title Case the model invented, keep tokens that look like names.
    private static func flattenModelCaps(_ text: String) -> String {
        var result = ""
        result.reserveCapacity(text.count)
        let characters = Array(text)
        var index = 0
        while index < characters.count {
            // Copy a protected span whole: flatten would lowercase the digits
            // inside it, and restore could not match the placeholder afterwards.
            if characters[index] == ProtectedSpans.open {
                let start = index
                while index < characters.count, characters[index] != ProtectedSpans.close {
                    index += 1
                }
                if index < characters.count { index += 1 }
                result.append(contentsOf: characters[start..<index])
                continue
            }
            let character = characters[index]
            if character.isLetter {
                let start = index
                index += 1
                while index < characters.count {
                    let next = characters[index]
                    if next.isLetter || next == "'" || next == "’" {
                        index += 1
                    } else {
                        break
                    }
                }
                result += softenToken(String(characters[start..<index]))
            } else {
                result.append(character)
                index += 1
            }
        }
        return result
    }

    private static func softenToken(_ token: String) -> String {
        if isPronounI(token) {
            let body = token.drop { !$0.isLetter }
            return String(token.dropLast(body.count)) + "I" + body.dropFirst()
        }
        let letters = token.filter(\.isLetter)
        if letters.isEmpty { return token }
        let hasLower = letters.contains { $0.isLowercase }
        let hasUpper = letters.contains { $0.isUppercase }
        if !hasLower && (2...4).contains(letters.count) { return token }
        if !hasLower && letters.count > 4 { return token.lowercased() }
        if hasLower && hasUpper {
            let body = token.drop { !$0.isLetter }
            guard let first = body.first else { return token }
            let titleCase = first.isUppercase && !body.dropFirst().contains(where: \.isUppercase)
            if !titleCase { return token }
        }
        return token.lowercased()
    }

    private static func isPronounI(_ token: String) -> Bool {
        let letters = token.filter(\.isLetter)
        guard let first = letters.first, first == "I" || first == "i" else { return false }
        let rest = String(letters.dropFirst()).lowercased()
        if rest.isEmpty { return true }
        let contracted = token.contains("'") || token.contains("’")
        return contracted && ["m", "ll", "d", "ve", "re", "s"].contains(rest)
    }

    private static func capitalizeSentenceStarts(
        _ text: String,
        punctuation: SentencePunctuation
    ) -> String {
        var result = ""
        var shouldCapitalize = true
        for character in text {
            if shouldCapitalize, character.isLetter {
                result += String(character).uppercased()
                shouldCapitalize = false
            } else {
                result.append(character)
            }
            if punctuation.terminators.contains(character) { shouldCapitalize = true }
        }
        return result
    }

    private static func casual(_ text: String, punctuation: SentencePunctuation) -> String {
        let capitalized = capitalizeSentenceStarts(text, punctuation: punctuation)
        guard !punctuation.terminator.isEmpty, !capitalized.hasSuffix("..") else { return capitalized }
        return capitalized.hasSuffix(punctuation.terminator)
            ? String(capitalized.dropLast(punctuation.terminator.count))
            : capitalized
    }

    private static func segments(
        _ text: String,
        punctuation: SentencePunctuation
    ) -> [String] {
        guard !text.isEmpty else { return [] }
        let characters = Array(text)
        var result: [String] = []
        var start = 0
        var index = 0
        while index < characters.count {
            let character = characters[index]
            let ellipsis = character == "."
                && (index > 0 && characters[index - 1] == "."
                    || index + 1 < characters.count && characters[index + 1] == ".")
            let next = index + 1 < characters.count ? characters[index + 1] : nil
            let boundary = punctuation.terminators.contains(character)
                && !ellipsis
                && (next == nil || next?.isWhitespace == true || punctuation.join.isEmpty)
            if boundary {
                result.append(String(characters[start...index]))
                start = index + 1
                while start < characters.count, characters[start].isWhitespace { start += 1 }
                index = start
            } else {
                index += 1
            }
        }
        if start < characters.count { result.append(String(characters[start...])) }
        return result.isEmpty ? [text] : result
    }

    private static func splitTerminator(
        _ sentence: String,
        punctuation: SentencePunctuation
    ) -> (body: String, terminator: String) {
        let body = sentence.trimmingCharacters(in: .whitespacesAndNewlines)
        guard let last = body.last,
              punctuation.terminators.contains(last),
              !(last == "." && body.hasSuffix(".."))
        else { return (body, "") }
        return (String(body.dropLast()), String(last))
    }

    private static func veryCasual(
        _ sentences: [String],
        punctuation: SentencePunctuation
    ) -> String {
        var parts: [String] = []
        for (index, sentence) in sentences.enumerated() {
            let split = splitTerminator(sentence, punctuation: punctuation)
            guard !split.body.isEmpty else { continue }
            if index == sentences.index(before: sentences.endIndex) {
                parts.append(split.body)
            } else if split.terminator.isEmpty || split.terminator == punctuation.terminator {
                parts.append(split.body + punctuation.separator)
            } else {
                parts.append(split.body + split.terminator)
            }
        }
        return parts.joined(separator: punctuation.join)
    }

    private static func excited(
        _ sentences: [String],
        punctuation: SentencePunctuation
    ) -> String {
        sentences.compactMap { sentence in
            let split = splitTerminator(sentence, punctuation: punctuation)
            guard !split.body.isEmpty else { return nil }
            if split.terminator == punctuation.question {
                return capitalizeSentenceStarts(split.body, punctuation: punctuation) + split.terminator
            }
            return capitalizeSentenceStarts(split.body, punctuation: punctuation) + punctuation.exclamation
        }
        .joined(separator: punctuation.join)
    }
}

File: android/app/src/main/java/com/vocahq/vocaphone/core/TranscriptStyler.kt
package com.vocahq.vocaphone.core

/**
 * Applies the same presentation-only contract as the gateway to a local
 * transcript. It deliberately never adds, removes, or substitutes words.
 *
 * Dropping a filler or inserting a missing sentence break would both break that
 * contract, which is why they are [TranscriptRepair]'s job and run before this
 * stage under a switch of their own.
 */
object TranscriptStyler {

    fun apply(text: String?, style: WritingStyle, language: String = "auto"): String {
        val source = text.orEmpty()
        if (style == WritingStyle.RAW) return source.trim()
        if (source.isBlank()) return ""

        val punctuation = SentencePunctuation.resolve(language, source)
        val spans = ProtectedSpans.mask(source)
        val normalized = normalizeSentenceTerminators(normalizeSpacing(spans.text), punctuation)
        // Whisper and Parakeet Title-Case content words ("the Keyboard Is Ready").
        // Flatten those before Clean/Formal/Casual so mid-sentence capitals are
        // not left as-is. Mixed-case names and ALL-CAPS acronyms stay.
        val flattened = when (style) {
            WritingStyle.RAW, WritingStyle.VERY_CASUAL -> normalized
            else -> flattenModelCaps(normalized)
        }
        val result = when (style) {
            WritingStyle.CLEAN -> ensureTerminator(flattened, punctuation)
            WritingStyle.FORMAL -> ensureTerminator(
                capitalizeSentenceStarts(flattened, punctuation),
                punctuation,
            )
            WritingStyle.CASUAL -> casual(flattened, punctuation)
            WritingStyle.VERY_CASUAL -> veryCasual(
                segments(normalized, punctuation),
                punctuation,
            )
            WritingStyle.EXCITED -> excited(
                segments(flattened, punctuation),
                punctuation,
            )
            WritingStyle.RAW -> normalized
        }
        val lowered = if (style == WritingStyle.VERY_CASUAL) {
            ProtectedSpans.mapOutsidePlaceholders(result) { it.lowercase() }
        } else {
            result
        }
        return spans.restore(lowered)
    }

    private fun normalizeSpacing(text: String): String = text
        .replace(Regex("\\s+"), " ")
        .trim()
        .replace(Regex("\\s+([.!?。！？।۔،,;:])"), "$1")

    /**
     * Models often emit an ASCII full stop for correctly decoded Hindi. Once
     * the script is known, normalize sentence boundaries while masked URLs,
     * decimals, abbreviations, and ellipses remain untouched.
     */
    private fun normalizeSentenceTerminators(
        text: String,
        punctuation: SentencePunctuation,
    ): String {
        if (punctuation.terminator != "।") return text
        return buildString(text.length) {
            text.forEachIndexed { index, character ->
                if (character != '.') {
                    append(character)
                    return@forEachIndexed
                }
                val previous = text.getOrNull(index - 1)
                val next = text.getOrNull(index + 1)
                val isEllipsis = previous == '.' || next == '.'
                val isSentenceBoundary = next == null || next.isWhitespace()
                append(if (isSentenceBoundary && !isEllipsis) '।' else character)
            }
        }
    }

    private fun ensureTerminator(text: String, punctuation: SentencePunctuation): String {
        if (text.isEmpty() || punctuation.terminator.isEmpty()) return text
        return if (text.last() in punctuation.terminators) text
        else text + punctuation.terminator
    }

    /**
     * Drop Title Case the model invented, keep tokens that look like names.
     *
     * "Keyboard" in the middle of a sentence becomes "keyboard". "VocaPhone",
     * "GraphQL", "iPhone", "NASA", and the pronoun "I" are left alone.
     */
    private fun flattenModelCaps(text: String): String {
        val result = StringBuilder(text.length)
        var index = 0
        while (index < text.length) {
            // Copy a protected span whole: flatten would lowercase the digits
            // inside it, and restore could not match the placeholder afterwards.
            if (text[index] == ProtectedSpans.OPEN) {
                val end = text.indexOf(ProtectedSpans.CLOSE, index)
                if (end >= 0) {
                    result.append(text, index, end + 1)
                    index = end + 1
                    continue
                }
            }
            if (text[index].isLetter()) {
                val start = index
                index++
                while (index < text.length &&
                    (text[index].isLetter() || text[index] == '\'' || text[index] == '’')
                ) {
                    index++
                }
                result.append(softenToken(text.substring(start, index)))
            } else {
                result.append(text[index])
                index++
            }
        }
        return result.toString()
    }

    private fun softenToken(token: String): String {
        if (isPronounI(token)) {
            val body = token.dropWhile { !it.isLetter() }
            return token.take(token.length - body.length) + "I" + body.drop(1)
        }
        val letters = token.filter { it.isLetter() }
        if (letters.isEmpty()) return token
        val hasLower = letters.any { it.isLowerCase() }
        val hasUpper = letters.any { it.isUpperCase() }
        // Short ALL-CAPS is an acronym (NASA, CPU). Longer shouts are the
        // model yelling a content word; flatten those.
        if (!hasLower && letters.length in 2..4) return token
        if (!hasLower && letters.length > 4) return token.lowercase()
        if (hasLower && hasUpper) {
            val body = token.dropWhile { !it.isLetter() }
            val titleCase = body.first().isUpperCase() && body.drop(1).none { it.isUpperCase() }
            if (!titleCase) return token
        }
        return token.lowercase()
    }

    /** "I", "I'm", "I’ve", "I'd", "I'll". Not "is" or "it". */
    private fun isPronounI(token: String): Boolean {
        val letters = token.filter { it.isLetter() }
        if (letters.isEmpty() || letters.first().lowercaseChar() != 'i') return false
        val rest = letters.drop(1).lowercase()
        if (rest.isEmpty()) return true
        val contracted = token.any { it == '\'' || it == '’' }
        return contracted && rest in setOf("m", "ll", "d", "ve", "re", "s")
    }

    private fun capitalizeSentenceStarts(
        text: String,
        punctuation: SentencePunctuation,
    ): String {
        val result = StringBuilder(text.length)
        var capitalize = true
        text.forEach { character ->
            if (capitalize && character.isLetter()) {
                result.append(character.uppercaseChar())
                capitalize = false
            } else {
                result.append(character)
            }
            if (character in punctuation.terminators) capitalize = true
        }
        return result.toString()
    }

    private fun casual(text: String, punctuation: SentencePunctuation): String {
        val capitalized = capitalizeSentenceStarts(text, punctuation)
        if (punctuation.terminator.isEmpty() || capitalized.endsWith("..")) return capitalized
        return if (capitalized.endsWith(punctuation.terminator)) {
            capitalized.dropLast(punctuation.terminator.length)
        } else {
            capitalized
        }
    }

    private fun segments(text: String, punctuation: SentencePunctuation): List<String> {
        if (text.isEmpty()) return emptyList()
        val result = mutableListOf<String>()
        var start = 0
        var index = 0
        while (index < text.length) {
            val character = text[index]
            val ellipsis = character == '.' && (
                index > 0 && text[index - 1] == '.' ||
                    index + 1 < text.length && text[index + 1] == '.'
                )
            val next = text.getOrNull(index + 1)
            val boundary = character in punctuation.terminators &&
                !ellipsis && (next == null || next.isWhitespace() || punctuation.join.isEmpty())
            if (boundary) {
                result += text.substring(start, index + 1)
                start = index + 1
                while (start < text.length && text[start].isWhitespace()) start++
                index = start
            } else {
                index++
            }
        }
        if (start < text.length) result += text.substring(start)
        return result.ifEmpty { listOf(text) }
    }

    private fun splitTerminator(
        sentence: String,
        punctuation: SentencePunctuation,
    ): Pair<String, String> {
        val body = sentence.trim()
        if (body.isEmpty()) return "" to ""
        val last = body.last()
        if (last in punctuation.terminators && !(last == '.' && body.endsWith(".."))) {
            return body.dropLast(1) to last.toString()
        }
        return body to ""
    }

    private fun veryCasual(
        sentences: List<String>,
        punctuation: SentencePunctuation,
    ): String {
        val parts = sentences.mapIndexedNotNull { index, sentence ->
            val (body, terminator) = splitTerminator(sentence, punctuation)
            if (body.isEmpty()) return@mapIndexedNotNull null
            if (index == sentences.lastIndex) body
            else if (terminator.isEmpty() || terminator == punctuation.terminator) {
                body + punctuation.separator
            } else {
                body + terminator
            }
        }
        return parts.joinToString(punctuation.join)
    }

    private fun excited(sentences: List<String>, punctuation: SentencePunctuation): String {
        val parts = sentences.mapNotNull { sentence ->
            val (body, terminator) = splitTerminator(sentence, punctuation)
            if (body.isEmpty()) null
            else if (terminator == punctuation.question) {
                capitalizeSentenceStarts(body, punctuation) + terminator
            } else {
                capitalizeSentenceStarts(body, punctuation) + punctuation.exclamation
            }
        }
        return parts.joinToString(punctuation.join)
    }
}
